src/client.ts

```diff
@@ -1022,6 +1024,24 @@ class ConnectionImpl implements Connection {
     }
   }
 
+  async transaction<T>(
+    action: () => Promise<T>,
+    options?: TransactionOptions
+  ): Promise<T> {
+    let result: T;
+    const transaction = new Transaction(this, options);
+    await transaction.start();
+    try {
+      result = await action();
```

Comment:
@elprans @RobertoPrevato We should pass the transaction object into the async closure, like this:


```
await con.transaction(async (tr) {
  ...
  if (something) {
    await conn.rollback();
  }
});
```

Reply:
Although we don't currently do that in asyncpg or in edgedb-python...

Reply:
@1st1; good point, I was also thinking that it would be nice to expose the transaction object (I also wrote this in my comment above a few days ago). I'm checking if TS supports defining overload by function argument's signature as it would be possible in C# or Java. Maybe we could manage the Transaction object only when the user passes a callback that doesn't expect any argument, and instead don't manage the transaction object when the user passes a callback that expects a transaction object as input.